src/workerd/server/container-client.c++

```diff
@@ -0,0 +1,406 @@
+    // To be safe we should inspect the container to get up to date mappings.
+    auto portMappings = kj::get<1>(co_await containerClient.inspectContainer());
+    auto maybeMappedPort = portMappings.find(containerPort);
+    if (maybeMappedPort == kj::none) {
+      throw JSG_KJ_EXCEPTION(DISCONNECTED, Error,
+          "connect(): Connection refused: container port not found. Make sure you exposed the port in your container definition.");
+    }
+    auto mappedPort = KJ_ASSERT_NONNULL(maybeMappedPort);
+
+    auto address =
+        co_await containerClient.network.parseAddress(kj::str(containerHost, ":", mappedPort));
+    auto connection = co_await address->connect();
+
+    auto upPipe = kj::newOneWayPipe();
+    auto upEnd = kj::mv(upPipe.in);
+    auto results = context.getResults();
+    results.setUp(containerClient.byteStreamFactory.kjToCapnp(kj::mv(upPipe.out)));
+    auto downEnd = containerClient.byteStreamFactory.capnpToKj(context.getParams().getDown());
+    pumpTask =
+        kj::joinPromisesFailFast(kj::arr(upEnd->pumpTo(*connection), connection->pumpTo(*downEnd)))
+            .ignoreResult()
+            .attach(kj::mv(upEnd), kj::mv(connection), kj::mv(downEnd));
+    co_return;
+  }
+
+ private:
+  ContainerClient& containerClient;
+  kj::String containerHost;
+  uint16_t containerPort;
+  kj::Maybe<kj::Promise<void>> pumpTask;
+};
+
+kj::Promise<ContainerClient::Response> ContainerClient::dockerApiRequest(
+    kj::HttpMethod method, kj::StringPtr endpoint, kj::Maybe<kj::StringPtr> body) {
+  kj::HttpHeaderTable headerTable;
+  auto address = co_await network.parseAddress(dockerPath);
+  auto connection = co_await address->connect();
+  auto httpClient = kj::newHttpClient(headerTable, *connection).attach(kj::mv(connection));
+  kj::HttpHeaders headers(headerTable);
+  headers.set(kj::HttpHeaderId::HOST, "localhost");
+
+  KJ_IF_SOME(requestBody, body) {
+    headers.set(kj::HttpHeaderId::CONTENT_TYPE, "application/json");
+    headers.set(kj::HttpHeaderId::CONTENT_LENGTH, kj::str(requestBody.size()));
+
+    auto req = httpClient->request(method, endpoint, headers, requestBody.size());
+    {
+      auto body = kj::mv(req.body);
+      co_await body->write(requestBody.asBytes());
+    }
+    auto response = co_await req.response;
+    auto result = co_await response.body->readAllText();
+    co_return Response{.statusCode = response.statusCode, .body = kj::mv(result)};
+  } else {
+    auto req = httpClient->request(method, endpoint, headers);
+    { auto body = kj::mv(req.body); }
+    auto response = co_await req.response;
+    auto result = co_await response.body->readAllText();
+    co_return Response{.statusCode = response.statusCode, .body = kj::mv(result)};
+  }
+}
+
+kj::Promise<kj::Tuple<bool, kj::HashMap<uint16_t, uint16_t>>> ContainerClient::inspectContainer() {
+  // Docker API: GET /containers/{id}/json
+  auto endpoint = kj::str("/containers/", containerName, "/json");
```

Comment:
Are we validating that `containerName` does not have any characters that would make the URL invalid? For instance, a `#` or `?` or `/` that is unescaped anywhere in the name.

Reply:
You're right. I just added a `kj::encodeUriComponent()` call to the constructor to fix this.